firmware/src/network_model.c

```diff
@@ -0,0 +1,837 @@
+    }
+  }
+  return -1;
+}
+
+/*
+ * @brief Check a request that contains an interface id.
+ * @returns -1 if the id was found and index was set, otherwise the size of the
+ * RDM response that was created.
+ */
+static int CheckInterfaceRequest(const RDMHeader *header,
+                                 const uint8_t *param_data,
+                                 unsigned int *index) {
+  if (header->param_data_length != sizeof(uint32_t)) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
+  }
+
+  unsigned int id = JoinUInt32(param_data[0], param_data[1],
+                               param_data[2], param_data[3]);
+  int found_index = LookupIndex(id);
+  if (found_index < 0) {
+    return RDMResponder_BuildNack(header, NR_DATA_OUT_OF_RANGE);
+  }
+  *index = found_index;
+  return -1;
+}
+
+/*
+ * @brief Simulate getting an DHCP address.
+ *
+ * This randomly fails and returns 0.0.0.0.
+ */
+static uint32_t GetDHCPAddress() {
+  if (Random_PseudoGet() % DHCP_FAILURE_RATIO == 0) {
+    // fail 1/3 of the time so we can test zeroconf
+    return IPV4_UNCONFIGURED;
+  }
+  return (10 << 24) + (Random_PseudoGet() & 0xffffff);
+}
+
+/*
+ * @brief Use a zeroconf address or fallback to unassigned.
+ */
+static void UseZeroconfOrUnassign(InterfaceState *interface) {
+  if (interface->current_zeroconf_mode) {
+    interface->current_ip = 0xa9fe0000 + (Random_PseudoGet() % 0xfeff);
+    interface->current_netmask = 16;
+    interface->config_source = CONFIG_SOURCE_ZEROCONF;
+  } else {
+    interface->config_source = CONFIG_SOURCE_NONE;
+    interface->current_ip = IPV4_UNCONFIGURED;
+    interface->current_netmask = 0;
+  }
+}
+
+static void ConfigureInterface(InterfaceState *interface) {
+  interface->current_dhcp_mode = interface->configured_dhcp_mode;
+  interface->current_zeroconf_mode = interface->configured_zeroconf_mode;
+
+  uint32_t dhcp_address;
+  if (interface->configured_ip) {
+    // Static IP, use that
+    interface->current_ip = interface->configured_ip;
+    interface->current_netmask = interface->configured_netmask;
+    interface->config_source = CONFIG_SOURCE_STATIC;
+  } else if (interface->configured_dhcp_mode &&
+             (dhcp_address = GetDHCPAddress())) {
+    interface->current_ip = dhcp_address;
+    interface->current_netmask = 8;
+    interface->config_source = CONFIG_SOURCE_DHCP;
+  } else {
+    UseZeroconfOrUnassign(interface);
+  }
+}
+
+// PID Handlers
+// ----------------------------------------------------------------------------
+int NetworkModel_GetListInterfaces(const RDMHeader *header,
+                                   UNUSED const uint8_t *param_data) {
+  if (header->param_data_length) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
+  }
+
+  ReturnUnlessUnicast(header);
+
+  unsigned int offset = sizeof(RDMHeader);
+  unsigned int i = 0;
+  for (; i < g_network_model.interface_count; i++) {
+    g_rdm_buffer[offset++] = UInt32Byte0(INTERFACE_DEFINITIONS[i].id);
```

Comment:
If this is actually in the right order, shouldn't the functions be named more logically, so 3 is MSB